wallet/wallet.go

```diff
@@ -0,0 +1,522 @@
+	}
+
+	tpoolSpent := make(map[types.Hash256]bool)
+	tpoolUtxos := make(map[types.Hash256]types.SiacoinElement)
+	for _, txn := range sw.cm.PoolTransactions() {
+		for _, sci := range txn.SiacoinInputs {
+			tpoolSpent[types.Hash256(sci.ParentID)] = true
+			delete(tpoolUtxos, types.Hash256(sci.ParentID))
+		}
+		for i, sco := range txn.SiacoinOutputs {
+			tpoolUtxos[types.Hash256(txn.SiacoinOutputID(i))] = types.SiacoinElement{
+				StateElement: types.StateElement{
+					ID: types.Hash256(types.SiacoinOutputID(txn.SiacoinOutputID(i))),
+				},
+				SiacoinOutput: sco,
+			}
+		}
+	}
+
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+	for _, sco := range outputs {
+		confirmed = confirmed.Add(sco.SiacoinOutput.Value)
+		if !sw.locked[sco.ID] && !tpoolSpent[sco.ID] {
+			spendable = spendable.Add(sco.SiacoinOutput.Value)
+		}
+	}
+
+	for _, sco := range tpoolUtxos {
+		unconfirmed = unconfirmed.Add(sco.SiacoinOutput.Value)
+	}
+	return
+}
+
+// Transactions returns a paginated list of transactions, ordered by block
+// height descending. If no more transactions are available, (nil, nil) is
+// returned.
+func (sw *SingleAddressWallet) Transactions(limit, offset int) ([]Transaction, error) {
+	return sw.store.Transactions(limit, offset)
+}
+
+// TransactionCount returns the total number of transactions in the wallet.
+func (sw *SingleAddressWallet) TransactionCount() (uint64, error) {
+	return sw.store.TransactionCount()
+}
+
+// FundTransaction adds siacoin inputs worth at least amount to the provided
+// transaction. If necessary, a change output will also be added. The inputs
+// will not be available to future calls to FundTransaction unless ReleaseInputs
+// is called.
+func (sw *SingleAddressWallet) FundTransaction(txn *types.Transaction, amount types.Currency) ([]types.Hash256, func(), error) {
+	if amount.IsZero() {
+		return nil, func() {}, nil
+	}
+
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
+	utxos, err := sw.store.UnspentSiacoinElements()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tpoolSpent := make(map[types.Hash256]bool)
+	tpoolUtxos := make(map[types.Hash256]types.SiacoinElement)
+	for _, txn := range sw.cm.PoolTransactions() {
+		for _, sci := range txn.SiacoinInputs {
+			tpoolSpent[types.Hash256(sci.ParentID)] = true
+			delete(tpoolUtxos, types.Hash256(sci.ParentID))
+		}
+		for i, sco := range txn.SiacoinOutputs {
+			tpoolUtxos[types.Hash256(txn.SiacoinOutputID(i))] = types.SiacoinElement{
+				StateElement: types.StateElement{
+					ID: types.Hash256(types.SiacoinOutputID(txn.SiacoinOutputID(i))),
+				},
+				SiacoinOutput: sco,
+			}
+		}
+	}
+
+	// remove locked and spent outputs
+	usableUTXOs := utxos[:0]
+	for _, sce := range utxos {
+		if sw.locked[sce.ID] || tpoolSpent[sce.ID] {
+			continue
+		}
+		usableUTXOs = append(usableUTXOs, sce)
+	}
+
+	// sort by value, descending
+	sort.Slice(usableUTXOs, func(i, j int) bool {
+		return usableUTXOs[i].SiacoinOutput.Value.Cmp(usableUTXOs[j].SiacoinOutput.Value) > 0
+	})
+
```

Comment:
Could we add an argument to `FundTransaction` called `useUnconfirmed` and then check here whether it's true and if so append unconfirmed transactions (also sorted by value) to `usableUTXOs`?

This makes a day and night difference when forming contracts on a fresh renter.